src/Components/Analyzers/src/ComponentParameterUsageAnalyzer.cs

```diff
@@ -0,0 +1,98 @@
+                {
+                    var assignedPropertyReferences = new List<IPropertyReferenceOperation>();
+                    startBlockContext.RegisterOperationAction(context =>
+                    {
+                        foreach (var operation in context.Operation.Children)
+                        {
+                            if (operation.Kind == OperationKind.PropertyReference)
+                            {
+                                assignedPropertyReferences.Add((IPropertyReferenceOperation)operation);
+                            }
+                        }
+                    }, OperationKind.SimpleAssignment, OperationKind.CompoundAssignment, OperationKind.CoalesceAssignment);
+
+                    startBlockContext.RegisterOperationBlockEndAction(context =>
+                    {
+                        for (var i = 0; i < assignedPropertyReferences.Count; i++)
+                        {
+                            var propertyReference = assignedPropertyReferences[i];
+                            var componentProperty = propertyReference.Member;
+                            var propertyAttributes = componentProperty.GetAttributes();
+
+                            if (!ComponentFacts.HasParameterAttribute(symbols, propertyAttributes))
+                            {
+                                // This is not a property reference that we care about, it is not decorated with [Parameter].
+                                continue;
+                            }
+
+                            var propertyLocationContainingType = componentProperty.ContainingType;
+                            if (!ComponentFacts.IsComponent(symbols, propertyLocationContainingType))
+                            {
+                                // Someone referenced a property as [Parameter] inside something that is not a component.
+                                continue;
+                            }
+
+                            var assignmentLocationContainingType = startBlockContext.OwningSymbol?.ContainingType;
+                            if (assignmentLocationContainingType == null)
+                            {
+                                // Assignment location has no containing type. Most likely we're operating on malformed code, don't try and validate.
+                                continue;
+                            }
+
+                            if (ComponentFacts.IsAssignableFrom(propertyLocationContainingType, assignmentLocationContainingType))
+                            {
+                                // The assignment is taking place within the components type hierarchy. This means the user is setting this in a supported
+                                // scenario.
+                                continue;
+                            }
+
+                            // At this point the user is referencing a component parameter outside of a component class.
```

Comment:
nitpick, outside of it's **declaring class**

The reason why I mention it, is that where people tend to break this rule is like this:

```razor
<ComponentA @ref="foo" />

@code {
    void OnClick()
    {
        foo.A = "Hiya!";
    }
}
```

This is inside of a component class 😆 and it's exactly what we want to prevent.